### PR TITLE
Add dali_data_type_vec 

### PR DESCRIPTION
**Pull request by majra20**
Signed-off-by: Rafal Maj <rmaj@nvidia.com>

#### Why we need this PR?
- Refactoring to improve type in documentation for numba_func schema defined in numba_func.cc

#### What happened in this PR?
*Fill relevant points, put NA otherwise. Replace anything inside []*
 - What solution was applied:
     *[ Added new type to type enum. ]*
 - Affected modules and functionalities:
     *[ Dali_data_type_vec enum is added. ]*
 - Key points relevant for the review:
     *[ NA ]*
 - Validation and testing:
     *[ Compile ]*
 - Documentation (including examples):
     *[ NA ]*

### CODE REVIEW

File: dali/python/nvidia/dali/types.py [hunk 1]
```diff
@@ -55,15 +55,17 @@ def _not_implemented(val):
         DALIDataType.INTERP_TYPE : ("nvidia.dali.types.DALIInterpType", lambda x: DALIInterpType(int(x))),
         DALIDataType.TENSOR_LAYOUT : (":ref:`layout str<layout_str_doc>`", lambda x: str(x)),
         DALIDataType.PYTHON_OBJECT : ("object", lambda x: x),
-        DALIDataType._TENSOR_LAYOUT_VEC : (":ref:`layout str<layout_str_doc>`", _to_list(lambda x: str(x)))
+        DALIDataType._TENSOR_LAYOUT_VEC : (":ref:`layout str<layout_str_doc>`", _to_list(lambda x: str(x))),
+        DALIDataType._DALI_DATA_TYPE_VEC : ("DALIDataType", _to_list(lambda x: str(DALIDataType.DATA_TYPE)))
```
Comment (jantonguirao) [inline dali/python/nvidia/dali/types.py:59]:
```suggestion
        DALIDataType._DALI_DATA_TYPE_VEC : ("nvidia.dali.types.DALIDataType", _to_list(lambda x: str(DALIDataType.DATA_TYPE)))
```
To match what is done in line 55

**Comment by majra20** [comment]
!build

File: dali/python/nvidia/dali/types.py (hunk 1 above)
Comment (klecki) [inline dali/python/nvidia/dali/types.py:59]: I'd say that this line is probably not needed, as it should go `_vector_types` and map _DALI_DATA_TYPE_VEC to a list of DALIDataType.DATA_TYPE

Comment (klecki) [reply dali/python/nvidia/dali/types.py:59]: If it is indeed necessary, it is wrong, as the encoding uses string and not int as the scalar DALIDataType does. It probably works due to not being used.

**Review by klecki — COMMENTED** [review]
Please remove the unnecessary mapping and adjust the Python enum value name, otherwise ok.

File: dali/operators/numba_function/numba_func.cc [hunk 2]
```diff
@@ -123,8 +123,8 @@ DALI_SCHEMA(NumbaFuncImpl)
   .AllowSequences()
   .SupportVolumetric()
   .AddArg("run_fn", R"code(Address of function to be invoked.)code", DALI_INT64)
-  .AddArg("out_types", R"code(Dali types of outputs.)code", DALI_INT_VEC)
-  .AddArg("in_types", R"code(Dali types of inputs.)code", DALI_INT_VEC)
+  .AddArg("out_types", R"code(Dali types of outputs.)code", DALI_DATA_TYPE_VEC)
+  .AddArg("in_types", R"code(Dali types of inputs.)code", DALI_DATA_TYPE_VEC)
   .AddArg("outs_ndim", R"code(Number of dimensions which outputs shapes should have.)code", DALI_INT_VEC)
   .AddArg("ins_ndim", R"code(Number of dimensions which inputs shapes should have.)code", DALI_INT_VEC)
   .AddOptionalArg<int>("setup_fn", R"code(Address of setup function setting shapes for outputs.
```
[the remark this answers is not in the document]
Comment (majra20) [reply dali/operators/numba_function/numba_func.cc:130]: The description is a bit different for run and setup functions. Am I able to override them?

File: include/dali/core/copy_vector_helper.h [hunk 3]
```diff
@@ -28,7 +28,7 @@ void copy_vector(std::vector<T> &out, const std::vector<S> &in) {
   out.reserve(in.size());
   out.clear();
   for (decltype(auto) v : in) {
-    out.emplace_back(v);
+    out.emplace_back(static_cast<T>(v));
```
[the remark this answers is not in the document]
Comment (majra20) [reply include/dali/core/copy_vector_helper.h:31]: Yes, it doesn't

File: dali/python/nvidia/dali/types.py (hunk 1 above)
Comment (majra20) [reply dali/python/nvidia/dali/types.py:59]: Other vector types go here as well. In `_vector_types` we are holding type of vector's element.

Comment (majra20) [reply dali/python/nvidia/dali/types.py:59]: That lambda is not used, but we still need to add this vector here.

**Comment by majra20** [comment]
!build

**Review by jantonguirao — APPROVED** [review]

**Review by klecki — APPROVED** [review]

File: dali/python/nvidia/dali/types.py (hunk 1 above)
Comment (klecki) [reply dali/python/nvidia/dali/types.py]: Right, my mistake.

**Pull request merged**